### ISSUE TITLE
Bundler Using Incorrect Version

### ISSUE THREAD

**Issue by nbibler** [dependabot/dependabot-core#2018]
I have a repository with a Gemfile and Gemfile.lock in it. The Gemfile.lock is generated by Bundler and contains the `BUNDLED WITH` information at the bottom:

```
BUNDLED WITH
   2.1.4
```

However, when Dependabot checks the repository, I'm getting `Dependabot can't resolve your Ruby dependency files` issues with:

```
Bundler::VersionConflict with message: Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    license_finder (~> 6.0) was resolved to 6.1.0, which depends on
      bundler (~> 2.1.0)

    rails (~> 6.0.0) was resolved to 6.0.2.1, which depends on
      bundler (>= 1.3.0)

  Current Bundler version:
    bundler (1.17.3)
```

Seemingly, the Dependabot runner is executing Bundler at version `1.17.3`, even though the Gemfile.lock identifies `2.1.4` as the version used to create the lock. That `BUNDLED WITH` [should be enough](https://bundler.io/guides/bundler_2_upgrade.html) for Bundler to know to use a 2.x version when executing, but seemingly it's not?

Is there a way to have Dependabot run using Bundler 2.x instead of 1.x that isn't documented somewhere?

**Comment by dpep** [comment]
any update or ways we can contribute to the Bundler 2 upgrade effort?

**Comment by feelepxyz** [comment]
👋  we've just launched support for bundler v2 in Dependabot Version and Security Updates. This should just work if your lockfile has been `BUNDLED WITH 2.x`. Let us know if this is working for you.

Dependabot Preview support is coming next week.

**Comment by jurre** [comment]
This should be resolved now 🎉 please let us know if you run into things

**Comment by dpep** [comment]
it's looking good on my end.  thanks for pushing the upgrade through!  🎉  🚀

**Comment by csalvato** [comment]
Hm I am seeing an issue using Bundler 2.1.4:

<img width="932" alt="Screen Shot 2021-04-22 at 4 51 18 PM" src="https://user-images.githubusercontent.com/1158152/115794283-f79ce100-a38a-11eb-859e-67e008fe5ad6.png">

It looks like Dependabot is using Bunder 2.2.15, but our lockfile was bundled with 2.1.4.

Upgrading to 2.2.15 will require a lift from our team.

Is there a way to  get this to work with Bundler 2.1.4?

Here's our `dependabot.yml`:

```yml
version: 2
registries:
  contribsys: # Define access for contribsys registry
    type: rubygems-server
    url: https://enterprise.contribsys.com/
    username: ${{secrets.SIDEKIQ_USERNAME}}
    password: ${{secrets.SIDEKIQ_PASSWORD}}
updates:
  - package-ecosystem: 'bundler'
    insecure-external-code-execution: allow
    registries:
      - contribsys
    open-pull-requests-limit: 100
    directory: '/'
    schedule:
      interval: 'monthly'
    labels:
      - 'dependencies'
      - 'dependabot'
      - 'ruby'

```

**Comment by feelepxyz** [comment]
@csalvato is your gemfile requiring bundler, e.g. `gem "bundler", "2.1.4"`?

If the project was bundled with it, it should work with any v2 version afaik but will complain if it's requiring a specific version that doesn't match the installed version. There might be something we can do to suppress these errors or not try and update bundler itself. Dynamically installing a custom version will be a pretty big lift.

**Comment by csalvato** [comment]
> @csalvato is your gemfile requiring bundler, e.g. `gem "bundler", "2.1.4"`?

Yes. On L9 of our `Gemfile`:

```rb
gem "bundler", "2.1.4"
```

And at the bottom of our `Gemfile.lock`:

```shell
BUNDLED WITH
   2.1.4
```

And when I look at the version installed by bundler locally:

```shell
$ bundle info bundler
  * bundler (2.1.4)
        Summary: The best way to manage your application's dependencies
        Homepage: https://bundler.io
        Path: /Users/csalvato/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/bundler-2.1.4
        Default Gem: yes
```

> If the project was bundled with it, it should work with any v2 version afaik but will complain if it's requiring a specific version that doesn't match the installed version.

It looks like we have bundled this project with v2.1.4 but are still seeing this error message. Any advice?

**Comment by feelepxyz** [comment]
> ```ruby
> ```ruby
> gem "bundler", "2.1.4"
> ```
> ```

Ah right, just to confirm, do you rely on `Bundler` in your project? You might be able to remove the `gem "bundler", "2.1.4"` line if you exec your code with bundler as this will be available without installing the gem.

**Comment by feelepxyz** [comment]
Alternatively relaxing the version constraint would also work with dependabot, `gem "bundler, "~> 2.1.4"`

**Comment by csalvato** [comment]
> Alternatively relaxing the version constraint would also work with dependabot, gem "bundler, "~> 2.1.4"

Thanks for your help @feelepxyz

This solution didn't resolve our problem.  We changed our Gemfile to read `gem "bundler", "~> 2.1.4"` to no avail.

It seems this had no effect, and now the check is continuing to hang:

<img width="880" alt="Screen Shot 2021-04-23 at 1 50 33 PM" src="https://user-images.githubusercontent.com/1158152/115923033-e27f8b00-a43a-11eb-9446-cdbb7e5bab88.png">

**Comment by csalvato** [comment]
It finally resolved, with the same error 🤔 

<img width="879" alt="Screen Shot 2021-04-23 at 2 02 26 PM" src="https://user-images.githubusercontent.com/1158152/115924124-8ae21f00-a43c-11eb-9ab5-31f9cb82ce2e.png">

**Comment by jurre** [comment]
Could you share a setup to reproduce that case @csalvato, that would help us debug what's going on

**Comment by csalvato** [comment]
@jurre Is this what you're looking for?

https://github.com/dependabot/dependabot-core/issues/2018#issuecomment-825233511

If not, can you clarify what you mean by "share a setup"?

**Comment by csalvato** [comment]
@jurre @feelepxyz Any insights into how I can resolve this?

**Comment by jurre** [comment]
> If not, can you clarify what you mean by "share a setup"?

Ideally a repo that replicates the failure, but the Gemfile and Gemfile.lock that go with the config you shared would already help a great deal

**Comment by csalvato** [comment]
We can't share the Gemfile + Gemfile.lock publicly for security and competitive reasons; is tere any way I can securely share that privately with you or the dependabot team directly?

**Comment by jurre** [comment]
> We can't share the Gemfile + Gemfile.lock publicly for security and competitive reasons; is tere any way I can securely share that privately with you or the dependabot team directly?

Yeah for sure, the easiest way would be to send an issue through GitHub support and reference this thread. One of the dependabot engineers will pick it up. But if you'd be able to set up a minimal Gemfile with most gems stripped out that reproduces the error, that'd be even easier.

**Comment by siegfault** [comment]
I've been able to replicate this issue with this gemfile: https://github.com/siegfault/simple_rails/blob/56922a8ad3a0a26bd3488af4953b7ef4eda62a66/Gemfile. It is the result of running `rails create --minimal`, and then adding `gem 'bundler', '~> 2.3.15'`. This results in
```
Dependabot can't resolve your Ruby dependency files
Dependabot failed to update your dependencies because there was an error resolving your Ruby dependency files.

Dependabot encountered the following error:

Bundler::VersionConflict with message: Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    bundler (~> 2.3.15)

  Current Bundler version:
    bundler (2.3.13)

Your bundle requires a different version of Bundler than the one you're running.
Install the necessary version with `gem install bundler:2.3.16` and rerun bundler using `run.rb _2.3.16_`
```
. Hope that helps!

**Comment by siegfault** [comment]
https://github.com/dependabot/dependabot-core/pull/5382 fixes the issue in my test repo as well as the one I originally had the problem in. For reference, here's the commit message I had worked up for 2.3.17 before that PR was opened:


    Fixes #2018. I am able to replicate the original issue using
    `LOCAL_GITHUB_ACCESS_TOKEN=redacted bin/dry-run.rb bundler siegfault/simple_rails`.
    
    I'm not positive whether the bundler version simply needs to be newer
    than the locked version, but I fear this may not work when 2.3.17 does
    not meet the requirements (e.g. `~> 2.1.4` in the original issue). It
    seems like we'd possibly need to install the correct version of bundler
    that matches the gemfile that we're about to parse, which seems a bit
    backwards, or at least difficult. Opening this to start the conversation
    at least.
